File: common/src/main/kotlin/net/impleri/playerskills/events/handlers/PlayerEvents.kt
package net.impleri.playerskills.events.handlers

import dev.architectury.event.EventResult
import dev.architectury.event.events.common.PlayerEvent
import net.impleri.playerskills.api.ItemRestrictions
import net.impleri.playerskills.server.NetHandler
import net.impleri.playerskills.skills.registry.Players
import net.impleri.playerskills.utils.PlayerSkillsLogger
import net.minecraft.server.level.ServerPlayer
import net.minecraft.world.entity.Entity
import net.minecraft.world.entity.player.Player
import net.minecraft.world.item.ItemStack

internal class PlayerEvents {
  fun registerEvents() {
    PlayerEvent.PLAYER_JOIN.register(PlayerEvent.PlayerJoin { onPlayerJoin(it) })
    PlayerEvent.PLAYER_QUIT.register(PlayerEvent.PlayerQuit { onPlayerQuit(it) })

    PlayerEvent.PICKUP_ITEM_PRE.register(
      PlayerEvent.PickupItemPredicate { player: Player, _: Entity, stack: ItemStack ->
        beforePlayerPickup(
          player,
          stack,
        )
      },
    )
  }

  private fun onPlayerJoin(player: ServerPlayer) {
    Players.open(player.uuid)
    NetHandler.syncPlayer(player, true)
  }

  private fun beforePlayerPickup(player: Player, stack: ItemStack): EventResult {
    val item = ItemRestrictions.getValue(stack)
    val itemName = ItemRestrictions.getName(item)

    if (!ItemRestrictions.canHold(player, item, null)) {
      PlayerSkillsLogger.ITEMS.debug("${player.name.string} cannot pickup $itemName")
      return EventResult.interruptFalse()
    }

    PlayerSkillsLogger.SKIPS.debug("${player.name.string} is going to pickup $itemName")
    return EventResult.pass()
  }

  private fun onPlayerQuit(player: ServerPlayer) {
    Players.close(player.uuid)
  }
}

File: common/src/main/scala/net/impleri/playerskills/server/bindings/PlayerEvents.scala
package net.impleri.playerskills.server.bindings

import dev.architectury.event.events.common.PlayerEvent
import dev.architectury.event.Event
import net.impleri.playerskills.facades.minecraft.Player
import net.impleri.playerskills.server.skills.PlayerRegistry
import net.impleri.playerskills.server.NetHandler

case class PlayerEvents(
  playerRegistry: PlayerRegistry = PlayerRegistry(),
  netHandler: NetHandler = NetHandler(),
  onPlayerJoin: Event[PlayerEvent.PlayerJoin] = PlayerEvent.PLAYER_JOIN,
  onPlayerQuit: Event[PlayerEvent.PlayerQuit] = PlayerEvent.PLAYER_QUIT,
) {
  def registerEvents(): Unit = {
    onPlayerJoin.register(p => playerJoin(Player(p)))
    onPlayerQuit.register(p => playerQuit(Player(p)))
  }

  private[bindings] def playerJoin(player: Player[_]): Unit = {
    playerRegistry.open(player.uuid)
    netHandler.syncPlayer(player)
  }

  private[bindings] def playerQuit(player: Player[_]): Unit = {
    playerRegistry.close(player.uuid)
  }
}
